### ISSUE TITLE
Autoupdate service hangs when `projection/content` is requested

### ISSUE THREAD

**Issue by FinnStutzenstein** [OpenSlides/openslides-autoupdate-service#254]
Even with #250, the error down below is still present:

```
Error parsing slide projection/6/content: fetching projection 6 from datastore: fetching data: getOrSet for keys `[projection/6/id projection/6/type projection/6/content_object_id projection/6/meeting_id projection/6/options]`: context canceled
```

There was a similar issue for projection 5. Those were a topic and a user. I started everything on master, opened the client, logged in, go the the first projector detail view, opened the agenda in a new tab and projected `A`. I did not receive an autoupdate there until I reload the projector-tab. With the reload the message above shows up in the logs.

**Comment by ostcar** [comment]
context canceled happens, when the connection is lost. Could it be a problem with the proxy?

**Comment by FinnStutzenstein** [comment]
It is triggered, if I reload the page manually, so no problem with the proxy.

**Comment by ostcar** [comment]
This also explains it. A reload stops the old request and creates a new one.

The error happens on the old request, when it is suddenly at an end.

This is still a bug. The error should not be logged.

I will also have a look at the place where the error happens. It could be in a (nearly) deadlock that only finishes, when the connection closes.

**Comment by ostcar** [comment]
Could you test this again with the current master (after #270)

**Comment by FinnStutzenstein** [comment]
No, it does not help since it does not solve the issue. Now the autoupdate hands for one seconds (I guess due to https://github.com/OpenSlides/openslides-autoupdate-service/pull/270/files#diff-bc85461a4330bf9a7f24943c17dea0b71e26621b9e3e57ed8b2f8f5d1d115e24R203). The one second is arbitrary and does not fix the issue's cause.  This is the error message:
```
2021/07/20 05:24:22 Error: calculating key projection/6/content: fetching projection 6 from datastore: fetching data: getOrSet for keys `[projection/6/id projection/6/type projection/6/content_object_id projection/6/meeting_id projection/6/options]`: waiting for pending key projection/6/type: context deadline exceeded
```
Also a test for this bug seems to be important.

**Comment by ostcar** [comment]
This issue is, that the autoupdate hangs when a bug happens on a projector. This is solved. Of cause, there is another bug in one of the slides. 

I think the bug is in the current list of speaker slide. Maybe @r-peschke can have a look at it?

**Comment by FinnStutzenstein** [comment]
> This issue is, that the autoupdate hangs when a bug happens on a projector.

But why? Can't you fix this? for the actual data in `projection/content` there is no difference, if there was an issue.

**Comment by ostcar** [comment]
I don't know this error. It seems, that there is a problem with the CLOS slide. @r-peschke implemented it. So maybe he can have a look?

**Comment by r-peschke** [comment]
I can have a look to catch this error. But
1. context deadline exceeded don't seem to be a bug in the projector code, it seems to be a bug caused by the environment
2. More important: I think we should catch ALL bugs, even from system. And, Oscar, I think that has to be done outside of the projector or the projector has to catch them all. But last one you didn't wanted.

**Comment by FinnStutzenstein** [comment]
I agree with @r-peschke here. We have two issues to distinguish. Lets focus on the hanging autoupdate when something else goes wrong **in this issue** and other bugs can be reported in different issues. There might be more bugs in the future (also in production) and these bugs should not cause the service to hang. It seems, that the service currently isn't reliable enough when errors, which can be handled (even by specs in the wiki) causing the service to not operate for anyone.

**Comment by ostcar** [comment]
I fixes this. Now, the service is reliable. The service does not hang anymore.

There is the pause of one second. But this is important. If two clients request the same projection/content at the same time, then it should be calculated only once. So one request has to wait until the first is done. But what is, when the first request has a bug and hangs? How long should the second request wait? I decided, that it should wait one second. But if you like, I can change this to 100 ms or another time. This time should not be to low or the first request will not succeed to calculate the field.

In conclusion: yes, there was a bug that the autoupdate service could hang if an error happend. Now, this is not possible anymore.

@r-peschke the projector code can now return an error. All errors are cached outside of the projector code.

**Comment by r-peschke** [comment]
@ostcar  What kind of error do you mean? 
Errors I get in the projector code, will be send to the client, breaking the error chain, in the regular response. 
Other environment/system errors I didn't catch and I assume you are referring to those ones.

**Comment by ostcar** [comment]
That is true. And that is what I mean. If an error in the projector code happens, the error chain is broken at the boarder to the autoupdate code. So an error in the projector does not break the service. I changed this a bit at #270. But the idea is the same.

There was a bug that this separation did not work. That was fixed with #270.

Now, if an error happens in a calculated field it does not break other connections. Other connections only wait for one second.

I don't think, that I am able to explain this better. If you like, you can call me.
